src/lib/address_bitset.cc

```diff
@@ -0,0 +1,90 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0. This product includes software
+// developed at Datadog (https://www.datadoghq.com/). Copyright 2021-Present
+// Datadog, Inc.
+#include "address_bitset.hpp"
+
+#include <algorithm>
+#include <functional>
+#include <unlikely.hpp>
+
+namespace ddprof {
+
+namespace {
+int most_significant_bit_pos(int n) {
+  if (n == 0)
+    return 0;
+  int msb = 0;
+  n = n / 2;
+  while (n != 0) {
+    n = n / 2;
+    msb++;
+  }
+  return msb;
+}
+} // namespace
+
+void AddressBitset::init(int sampling_period, unsigned max_addresses) {
+  _lower_bits_ignored = most_significant_bit_pos(sampling_period);
+  // we avoid ignoring too many of the lower bits
+  _lower_bits_ignored = std::min<int>(_k_max_bits_ignored, _lower_bits_ignored);
+  if (_address_bitset) {
+    _address_bitset.reset();
+  }
+  _nb_bits = max_addresses;
+  _k_nb_elements = (_nb_bits) / (_nb_bits_per_elt);
+  if (_nb_bits) {
+    _nb_bits_mask = _nb_bits - 1;
+    _address_bitset = std::make_unique<std::atomic<uint64_t>[]>(_k_nb_elements);
+  }
+}
+
+bool AddressBitset::set(uintptr_t addr) {
+  uint64_t hash_addr = remove_lower_bits(addr);
```

Comment:
I think using a standard hash has some merit.

Reply:
This strategy is also interesting considering we don't expect a lot of interesting consecutive addresses for live heap.